sql/parser/type_check.go

```diff
@@ -0,0 +1,388 @@
+
+	case *FuncExpr:
+		return typeCheckFuncExpr(t)
+
+	case *CaseExpr:
+		return typeCheckCaseExpr(t)
+
+	case *CastExpr:
+		return typeCheckCastExpr(t)
+	}
+
+	return nil, util.Errorf("eval: unexpected expression: %T", expr)
+}
+
+func typeCheckBooleanExprs(exprs ...Expr) (reflect.Type, error) {
+	for _, expr := range exprs {
+		exprType, err := TypeCheckExpr(expr)
+		if err != nil {
+			return nil, err
+		}
+		if exprType != boolType {
+			return nil, fmt.Errorf("incompatible AND argument type %s", typeToValue[exprType].Type())
+		}
+	}
+	return boolType, nil
+}
+
+// TODO(tamird): reduce duplication with evalComparisonExpr.
+func typeCheckComparisonExpr(expr *ComparisonExpr) (reflect.Type, error) {
+	leftType, err := TypeCheckExpr(expr.Left)
+	if err != nil {
+		return nil, err
+	}
+	rightType, err := TypeCheckExpr(expr.Right)
+	if err != nil {
+		return nil, err
+	}
+
+	return typeCheckComparisonOp(expr.Operator, leftType, rightType)
+}
+
+func typeCheckComparisonOp(op ComparisonOp, leftType, rightType reflect.Type) (reflect.Type, error) {
+	switch op {
+	case NE:
+		// NE(left, right) is implemented as !EQ(left, right).
+		op = EQ
+	case GT:
+		// GT(left, right) is implemented as LT(right, left)
+		op = LT
+		leftType, rightType = rightType, leftType
+	case GE:
+		// GE(left, right) is implemented as LE(right, left)
+		op = LE
+		leftType, rightType = rightType, leftType
+	case NotIn:
+		// NotIn(left, right) is implemented as !IN(left, right)
+		op = In
+	}
+
+	// TODO(pmattis): Memoize the cmpOps lookup as we've done for unaryOps and
+	// binOps.
+	if f, ok := cmpOps[cmpArgs{op, leftType, rightType}]; ok {
+		return reflect.TypeOf(f).Out(0), nil
+	}
+
+	switch op {
+	case Like, NotLike:
+		return nil, util.Errorf("TODO(pmattis): unsupported comparison operator: %s", op)
+	}
+
+	return nil, fmt.Errorf("unsupported comparison operator: <%s> %s <%s>",
+		typeToValue[leftType].Type(), op, typeToValue[rightType].Type())
+}
+
+func typeCheckRangeCond(expr *RangeCond) (reflect.Type, error) {
+	var newExpr Expr = &AndExpr{
+		Left: &ComparisonExpr{
+			Operator: GE,
+			Left:     expr.Left,
+			Right:    expr.From,
+		},
+		Right: &ComparisonExpr{
+			Operator: LE,
+			Left:     expr.Left,
+			Right:    expr.To,
+		},
+	}
+
+	if expr.Not {
+		newExpr = &NotExpr{
+			Expr: newExpr,
+		}
+	}
+
+	return TypeCheckExpr(newExpr)
+}
+
+func typeCheckBinaryExpr(expr *BinaryExpr) (reflect.Type, error) {
+	leftType, err := TypeCheckExpr(expr.Left)
+	if err != nil {
+		return nil, err
+	}
+	rightType, err := TypeCheckExpr(expr.Right)
+	if err != nil {
+		return nil, err
+	}
+	expr.ltype = leftType
+	expr.rtype = rightType
+	left := typeToValue[leftType]
+	right := typeToValue[rightType]
+
+	var ok bool
+	if expr.fn, ok = binOps[binArgs{expr.Operator, leftType, rightType}]; !ok {
+		return nil, fmt.Errorf("unsupported binary operator: <%s> %s <%s>",
+			left.Type(), expr.Operator, right.Type())
+	}
+	d, err := expr.fn(left, right)
```

Comment:
Rather the calling the function, perhaps we should expand `binOps` to hold both the function and the return value type. I think this would remove the need for the `dInt` hack.